### PR TITLE
go/cmd/fixgenesis: Add default beacon and governance parameters

### PR DESCRIPTION
**Pull request by ptrus**
Fixes: #3728

TODO:
- [x] see todo's in code
- [x] remove entities not passing stake claims (done, but blocked by: #3754)

### CODE REVIEW

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go [hunk 1]
```diff
@@ -301,27 +324,57 @@ func updateGenesisDoc(oldDoc *oldDocument) (*genesis.Document, error) {
...
+	oldEpochInterval := oldDoc.Epochtime.Parameters.Interval
+	newDoc.Beacon.Base = beacon.EpochTime(oldDoc.Height / oldEpochInterval)
+	newDoc.Beacon.Parameters.Backend = beacon.BackendPVSS
+	newDoc.Beacon.Parameters.PVSSParameters = &beacon.PVSSParameters{
+		// TODO: These were taken from E2E tests.
+		CommitInterval:  oldEpochInterval / 2,
+		RevealInterval:  (oldEpochInterval / 2) - 4,
+		TransitionDelay: 4,
+		// TODO: sensible defaults? Compute based on old doc validator set?
+		Threshold:    15,
```
Comment (kostko) [inline go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:347]: This could probably be 10.

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go [hunk 2]
```diff
@@ -301,27 +324,57 @@ func updateGenesisDoc(oldDoc *oldDocument) (*genesis.Document, error) {
...
-		Consensus:  oldDoc.Consensus,
 		HaltEpoch:  oldDoc.HaltEpoch,
 		ExtraData:  oldDoc.ExtraData,
 	}
 
+	// Consensus.
+	newDoc.Consensus.Backend = oldDoc.Consensus.Backend
+	newDoc.Consensus.Parameters = oldDoc.Consensus.Parameters.Parameters
+	// TODO: different constant?
+	newDoc.Consensus.Parameters.MaxEvidenceSize = 100 * uint64(oldDoc.Consensus.Parameters.MaxEvidenceNum)
```
Comment (kostko) [inline go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:335]: Probably should be 1024 to be safe.

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go [hunk 3]
```diff
@@ -301,27 +324,57 @@ func updateGenesisDoc(oldDoc *oldDocument) (*genesis.Document, error) {
...
+		Threshold:    15,
+		Participants: 20,
+	}
+
+	// Roothash.
 	newDoc.RootHash.Parameters.MaxRuntimeMessages = 32
 
+	// Governance.
+	newDoc.Governance.Parameters = governance.ConsensusParameters{
+		MinProposalDeposit: *quantity.NewFromUint64(100000000000), // TODO: more?
```
Comment (kostko) [inline go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:356]: Yeah this should probably be a bit more, like `10_000_000_000_000`?

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go [hunk 4]
```diff
@@ -301,27 +324,57 @@ func updateGenesisDoc(oldDoc *oldDocument) (*genesis.Document, error) {
...
+		CommitInterval:  oldEpochInterval / 2,
+		RevealInterval:  (oldEpochInterval / 2) - 4,
+		TransitionDelay: 4,
+		// TODO: sensible defaults? Compute based on old doc validator set?
+		Threshold:    15,
+		Participants: 20,
+	}
+
+	// Roothash.
 	newDoc.RootHash.Parameters.MaxRuntimeMessages = 32
```
Comment (kostko) [inline go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:352]: This should be increased to 256.

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go (hunk 2 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:335]: Just 1024 or `1024 * existing_MaxEvidenceNum`?

**Comment by Yawning** [comment]
So, I'm moderately against automagically fixing up the beacon parameters because it is deployment specific, and ostensibly oasis-core is deployment agnostic.  Realistically there is one deployment that matters so if people are ok with a command in a supposedly generic codebase using values that are specific to one deployed instance of the codebase, then the values seem ok (though in that case the reveal interval could probably be shortened).

**Comment by kostko** [comment]
Yeah that's always a problem. We could also move the mainnet specific tool to a separate repo.

> though in that case the reveal interval could probably be shortened

Any suggestions? So it needs to be enough for everyone to have a proposal slot.

**Comment by tjanez** [comment]
Since this is a debug tool, I would vote for removing the hard failure in case the sanity checks don't pass:

https://github.com/oasisprotocol/oasis-core/blob/3c302bdad91ead313f4039169a4837c94ba60c8d/go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go#L189-L195

With something like:

```go
	// Validate the new genesis document.
	if err = newDoc.SanityCheck(); err != nil {
		logger.Warn("new genesis document sanity check failed",
			"err", err,
		)
	}
```

This will still warn the user but also give him a (potentially broken) new genesis file which he can fix manually later.

Or not at all if the failure is something like:
```
level=warn module=cmd/debug/fix-genesis caller=fixgenesis.go:214 msg="new genesis document sanity check failed" err="registry: sanity check failed: one or more unsafe debug flags set"
```
which is expected in certain use cases.

**Comment by ptrus** [comment]
> Since this is a debug tool, I would vote for removing the hard failure in case the sanity checks don't pass:

Yeah good point, agreed. Will update to only warn.

**Review by kostko — APPROVED** [review]

**Pull request merged**